### ISSUE TITLE
Add password recovery

### ISSUE THREAD

**Issue by ZeRego** [lightdash/lightdash#1016]
### Description: What is it?

Add self served password recovery mechanism.

### Problem: What problem does this solve?

Users can't recover / change their password if they forget about it.

### Success: How will we know if we’ve solved this problem?

Users won't need to reach out for support.

### Audience: Who are we building this for?

 
Everyone

### What: Roughly, what does this look like in the product?

- In the login page we should have a button 'I forgot my password'. 
- Have a form where the user introduce their email 
- Send email with link to change password ( link has a single use token for security that expires in 1 day )
- Have a form where the user sets the new password

**Comment by owlas** [comment]
# Implementation

This requires Lightdash to start sending emails.

At minimum the Lightdash admin will need to provide the following variables by choosing an email provider (e.g. postmark, mailgun, Sendgrid,..): Email SMTP host, port, user, password, tls/ssl settings.

**Sending emails from Lightdash backend**

Lightdash needs to send emails

- [ ] Backend is provided with environment variables `EMAIL_SMTP_HOST`, `EMAIL_SMTP_PORT`, ...
- [ ] Backend uses an [email client](https://nodemailer.com/about/) to send emails using provided SMTP settings.

**Verifying emails**

Users **must** prove they own the email before we start sending password reset messages. So we add a new flag "verified" to the `emails` table. 

- [ ] Database has a new column on `emails` table called `is_verified`
- [ ] Frontend has option to "verify email"
- [ ] Backend sends an email containing a verification link, including a long secret code
- [ ] Database has a new table `email_verification_codes` containing `code,expires_at,email_id`
- [ ] When the user clicks the link, we check the `email_verification_codes` for a matching code and if it matches we verify the email. This is very similar to how we use invite codes.

**Password reset**

Users want to reset their password using an email flow.

- [ ] IF the backend has email configured (through the `EMAIL_SMTP...` environment variables) - then show a "forgot password?" option
- [ ] User fills in email in the frontend. Sees "Thanks we've sent a password reset email" (even if the account doesn't exist)
- [ ] Backend checks the database for a verified email matching the request
- [ ] Database has a new table `password_reset_links`, containing `code,expires_at,email_id`
- [ ] The reset code is sent to the verified email
- [ ] User opens emails and clicks the link
- [ ] User has option to reset their password

**Comment by ZeRego** [comment]
@owlas

> Users must prove they own the email before we start sending password reset messages.
A user that hasn't validated his email won't have a way to recover his account ? Why don't we send an email either way ? 

Other tasks:
[ ] Send verify email when user creates an account 
[ ] Add 'verify email' button in the profile section

**Comment by owlas** [comment]
> Why don't we send an email either way ?

It's for security. If you haven't proved that they own the email, then you're potentially giving some other user (maybe they mistyped their email, or it's an attack) the full power to reset that user's password. 

It's dangerous to send a password reset email if we haven't verified that the user actually owns the specific email address.

**Comment by ZeRego** [comment]
> It's for security. If you haven't proved that they own the email, then you're potentially giving some other user (maybe they mistyped their email, or it's an attack) the full power to reset that user's password.

Scenarios:
1. The user used a dummy email that he doesn't have access to. 

The user knows the recovery email will be sent to the account email which he doesn't have access to. If he submits the form, the "wrong" person will have access to the Lightdash account and in this scenario it's the user fault for creating an account that doesn't belong to him and to go forward with the recovery flow.
The only way to regain access  is for another user to send another invite link.

2. The user mistyped their email on user creation.  

To recover the password they type the correct email 'jose@example.com' but the user email is josee@example.com.
The UI will show that we sent the email but we didn't since we don't find a match for the email in our DB.
In case it knows the mistyped email, it will be similar to scenario 1, the only alternative is for another user to send another invite link.

3. The user used the correct email but forgot password.

The user is able to reset his password.

**Comment by ZeRego** [comment]
@PriPatel thoughts on UI and email ? Currently I only have a plain text version, we add a html version.

<a href="https://www.loom.com/share/39931ff4af2b4dd098bf200e34446fef">
    <p>Lightdash - Reset password - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/39931ff4af2b4dd098bf200e34446fef-with-play.gif">
  </a>

**Comment by hamzahc1** [comment]
You could simplify the text above the box - suggestion: 
Title: "Reset Password"
Subtitle: "Enter your email to reset your password"

**Comment by PriPatel** [comment]
@ZeRego Cool!! A couple of points/questions:

- +1 to simplifying text as per Hamzah's comment above!
- I take it that it is too much work to show the user a different screen after they have successfully entered their email address? Right now we just show a toast and that is ok but it would be much nicer to have a more obvious confirmation. This way, it is much clearer to the user what is happening. We can also have a clearer message about. the. 'resend email' button and ask the user to double check for typos if they haven't received anything yet etc. I can send a sketch tomorrow if you think this is doable. 
- I will come up with the copy for the plain text version but also a very basic sketch for the html version as a nice-to-have (unless you think the html version is something that will happen much later down the line!)

Thanks!

**Comment by ZeRego** [comment]
@PriPatel All those changes should be quick to do tomorrow 👍  thank you 🙏

**Comment by owlas** [comment]
@ZeRego and I discussed security of verifying emails

We agreed that email verification is important. We'll merge the password-reset flow and prioritise a good UX for email verification.

**Comment by PriPatel** [comment]
@ZeRego 
Here are two flows. One for **requesting the password link**, and one for **resetting the password**.
https://www.figma.com/file/Mhuwki1Bj1akQIwPyr1gWw/Lightdash-product?node-id=166%3A24874

- Please ignore the styling in the files as Ive just hacked it together quickly - it is mainly to inform the flow and the microcopy. Let me know if you have any questions or feedback :)

**Comment by ZeRego** [comment]
@PriPatel @nathaliabj I don't think we have that blue link/button anywhere in the app currently. Do we want to use the generic button component and just change the text color ? If so, it will have a padding all around. This is going to cause a small misalignment with the rest of the form.
If we want a brand new link/button component how does it look like on hover ?

**Comment by PriPatel** [comment]
@ZeRego @nathaliabj 
The button isn't going to work because of the way that they are aligned but I was hoping that it would be just an a href link [as shown in the blueprint documentation here. ](https://blueprintjs.com/docs/#core/typography.running-text)

Does this work? If not, then lets keep the button you have chosen and move the button from the far left so that it is next to the button on the right, like the original design: 
<img width="182" alt="Screenshot 2022-01-11 at 20 02 20" src="https://user-images.githubusercontent.com/12776522/148977700-449ba6f9-1d40-4b20-92f2-f2f53ee5a9ed.png">

**Comment by ZeRego** [comment]
@PriPatel You are correct, we can also use a `<a>` element ( with some extra logic to prevent a page refresh similar to <LinkButton/> )

**Comment by PriPatel** [comment]
@ZeRego cool! I think the appearance of a link and underline on hover is fine as these are supposed to be secondary actions. 
You're right in that they are buttons because the user is carrying out an action but yeah, the ideal button doesn't currently exist atm 😅
